packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts

```diff
@@ -0,0 +1,117 @@
+  limit: number;
+  datasourceUid: string;
+}>;
+
+export function isSuggestionsIncompleteEvent(e: Event): e is SuggestionsIncompleteEvent {
+  return (
+    e.type === CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT &&
+    'detail' in e &&
+    typeof e.detail === 'object' &&
+    e.detail !== null &&
+    'limit' in e.detail &&
+    'datasourceUid' in e.detail
+  );
+}
+
+interface Metric {
+  name: string;
+  help: string;
+  type: string;
+}
+
+export interface DataProviderParams {
+  languageProvider: PromQlLanguageProvider;
+  historyProvider: Array<HistoryItem<PromQuery>>;
+}
+
+export class DataProvider {
+  readonly languageProvider: PromQlLanguageProvider;
+  readonly historyProvider: Array<HistoryItem<PromQuery>>;
+  readonly getSeriesLabels: typeof this.languageProvider.getSeriesLabels;
+  readonly getSeriesValues: typeof this.languageProvider.getSeriesValues;
+  readonly getAllLabelNames: typeof this.languageProvider.getLabelKeys;
+  readonly getLabelValues: typeof this.languageProvider.getLabelValues;
+  readonly metricNamesSuggestionLimit: number;
+  /**
+   * The text that's been typed so far within the current {@link Monaco.Range | Range}.
+   *
+   * @remarks
+   * This is useful with fuzzy searching items to provide as Monaco autocomplete suggestions.
+   */
+  private inputInRange: string;
+  private suggestionsIncomplete: boolean;
+
+  constructor(params: DataProviderParams) {
+    this.languageProvider = params.languageProvider;
+    this.historyProvider = params.historyProvider;
+    this.inputInRange = '';
+    this.metricNamesSuggestionLimit = this.languageProvider.datasource.metricNamesAutocompleteSuggestionLimit;
+    this.suggestionsIncomplete = false;
+    this.getSeriesLabels = this.languageProvider.getSeriesLabels.bind(this.languageProvider);
+    this.getSeriesValues = this.languageProvider.getSeriesValues.bind(this.languageProvider);
+    this.getAllLabelNames = this.languageProvider.getLabelKeys.bind(this.languageProvider);
+    this.getLabelValues = this.languageProvider.getLabelValues.bind(this.languageProvider);
+  }
+
+  getHistory(): string[] {
+    return this.historyProvider.map((h) => h.query.expr).filter((expr) => expr !== undefined);
+  }
+
+  getAllMetricNames(): string[] {
+    return this.languageProvider.metrics;
+  }
+
+  metricNamesToMetrics(metricNames: string[]): Metric[] {
+    const { metricsMetadata } = this.languageProvider;
+    const result: Metric[] = metricNames.map((m) => {
+      const metaItem = metricsMetadata?.[m];
+      return {
+        name: m,
+        help: metaItem?.help ?? '',
+        type: metaItem?.type ?? '',
+      };
+    });
+
+    return result;
+  }
+
+  private enableAutocompleteSuggestionsUpdate(): void {
+    this.suggestionsIncomplete = true;
+    dispatchEvent(
+      new CustomEvent(CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT, {
+        detail: { limit: this.metricNamesSuggestionLimit, datasourceUid: this.languageProvider.datasource.uid },
+      })
+    );
+  }
```

Comment:
Running locally, I'm having trouble seeing the listener catch this event in the component `PromQueryCodeEditorAutocompleteInfo.tsx` that shows the warning. I might be missing something, but having trouble following the logic to get the warning to show up

Reply:
Sorry about that, @bohandley! There was a regression. Fixed now, by https://github.com/grafana/grafana/pull/85396/commits/49f8ce5bcfd4aef528a8c490c32e06b5b8373ce7:

https://github.com/grafana/grafana/assets/5732000/97a52405-c9ff-442c-ab61-7958801bea3d

Reply:
Beautiful! Works like a charm. Fantastic work on this!!!